Add more ways to seek to sample points

OliBomby opened the pull request:
This feature unlocks a powerful UX that allows users to quickly edit multiple samples by double-clicking and using hotkeys, without having to move the mouse cursor.

OliBomby committed:
Allow seeking to sample point on double-click

OliBomby committed:
Add test coverage

bdach commented:
I'm gonna be frank, I'm not sure I see what is "powerful" in this.

- For adding additional sounds, W~T keys are faster.
- For changing bank of all sounds at once, W~R keys are faster.
- For setting volume or setting separate normal/addition banks, clicking once is faster than clicking twice.

Is it the seek that's supposed to be making this "powerful"? Then let's maybe just have the popover seek every time?

In general I'm not sure this is going to be winning many favours. I've already heard from multiple feedback sources that the sample pieces on the timeline are difficult to work with when the timeline has high object density.

I'm honestly thinking we will inevitably have to go to the stable model of having clickable slider nodes themselves on playfield and/or timeline. That said I would want @peppy buy-in on this, because it has severe implications, such as killing the "all rulesets get a single unchanging implementation of timeline" premise, but I personally think that premise is dead in the water already if we wanna support everyone's use cases properly (case in point: scrolling rulesets and recurring complaints that you can't see scroll speed changes on timeline - which is because we moved them to timing screen).

OliBomby commented:
> I'm gonna be frank, I'm not sure I see what is "powerful" in this.
> 
> * For adding additional sounds, W~T keys are faster.
> * For changing bank of all sounds at once, W~R keys are faster.

It's powerful in combination with these features. Mappers usually have a specific pattern of hitsounds in mind and need to apply different hitsounds to each object. The automatic seek allows them to: click - press keys for hs - click - press keys for hs - ...
All without moving the mouse if its on the same snap. It optimizes the flow ever so slightly.

> Is it the seek that's supposed to be making this "powerful"? Then let's maybe just have the popover seek every time?

I think this is also acceptable UX.

> I'm honestly thinking we will inevitably have to go to the stable model of having clickable slider nodes themselves on playfield and/or timeline.

Does this imply removing the sample point pieces entirely? Would be good to know before I try redesigning them again to hopefully make them bearable to use.

I'm not against going to the stable model, it works pretty well and we can potentially have a separate hitsounding tab for more advanced hitsounds or modding.

> it has severe implications, such as killing the "all rulesets get a single unchanging implementation of timeline" premise, but I personally think that premise is dead in the water already if we wanna support everyone's use cases properly (case in point: scrolling rulesets and recurring complaints that you can't see scroll speed changes on timeline - which is because we moved them to timing screen).

I'm also in favour of allowing ruleset-specific implementations of the timeline. I ran into issues with this back when https://github.com/ppy/osu/pull/21566

bdach commented:
> Does this imply removing the sample point pieces entirely? Would be good to know before I try redesigning them again to hopefully make them bearable to use.

I'm honestly not sure. I haven't really tried to come up with what the alternative would be yet. Maybe they would stay, but as display-only and not be directly interactable.

peppy commented:
I don't know, this feels really awkward with its current implementation:

https://github.com/ppy/osu/assets/191335/1db68486-2143-4593-b3cb-75ffaf6e2614

Rather than this, would it not be better to have something like `Alt`+`Left`/`Right` skip to the next object (generally in editor) and making `Shift`+`Alt`+`Left`/`Right` bring up the samples edit box or similar?

peppy commented:
As for a better path forward, I think there should be a way to expand the timeline out and reveal a hitsounding section (can take up as much height as required) or even a separate hitsounding mode as suggested by @OliBomby where using left/right arrows seeks between hitsound nodes (similar to keyframe skipping in a video editor).

bdach commented:
> I think there should be a way to expand the timeline out and reveal a hitsounding section (can take up as much height as required) or even a separate hitsounding mode

I would generally be on board with this ([have floated the same previously](https://github.com/ppy/osu/pull/23445#issuecomment-1542610541)) but I'm not super sure all users will be happy with just that. Some are used to being able to seamlessly hitsound *while* mapping and as such, these types of modality changes where you need to go into and out of a special hitsounding mode are not likely to work super well for them.

OliBomby committed:
alt left/right or scroll to seek to neighbouring hit objects

OliBomby committed:
shift alt seek to open next sample edit popover

OliBomby committed:
always seek on click

OliBomby committed:
fix naming violation

OliBomby committed:
update tests

OliBomby commented:
I've implemented both of your suggestions.

- It now seeks on click instead of double-click. I think this makes it a lot less jank.
- I've added `alt` + `left`/`right` to skip to the next object.
- I've added `shift` + `alt` + `left`/`right` to skip to the next sample point and automatically open it. This will be very useful for accessing those very densely packed sample points.

OliBomby committed:
clarify logic

peppy commented:
I'm not 100% sure about the auto scroll on single click.

peppy commented on osu.Game/Screens/Edit/Editor.cs:
```diff
@@ -1064,8 +1065,80 @@ private void seekControlPoint(int direction)
+
         private void seek(UIEvent e, int direction)
         {
+            if (e.AltPressed)
```
To make this work, I need to apply this patch:

```diff
diff --git a/osu.Game/Overlays/Volume/VolumeControlReceptor.cs b/osu.Game/Overlays/Volume/VolumeControlReceptor.cs
index 4ddbc9dd48..2e8d86d4c7 100644
--- a/osu.Game/Overlays/Volume/VolumeControlReceptor.cs
+++ b/osu.Game/Overlays/Volume/VolumeControlReceptor.cs
@@ -23,15 +23,15 @@ public bool OnPressed(KeyBindingPressEvent<GlobalAction> e)
             {
                 case GlobalAction.DecreaseVolume:
                 case GlobalAction.IncreaseVolume:
-                    ActionRequested?.Invoke(e.Action);
-                    return true;
+                    return ActionRequested?.Invoke(e.Action) == true;
 
                 case GlobalAction.ToggleMute:
                 case GlobalAction.NextVolumeMeter:
                 case GlobalAction.PreviousVolumeMeter:
                     if (!e.Repeat)
-                        ActionRequested?.Invoke(e.Action);
-                    return true;
+                        return ActionRequested?.Invoke(e.Action) == true;
+
+                    return false;
             }
 
             return false;
diff --git a/osu.Game/Overlays/VolumeOverlay.cs b/osu.Game/Overlays/VolumeOverlay.cs
index 5470c70400..fa6e797c9c 100644
--- a/osu.Game/Overlays/VolumeOverlay.cs
+++ b/osu.Game/Overlays/VolumeOverlay.cs
@@ -120,14 +120,18 @@ public bool Adjust(GlobalAction action, float amount = 1, bool isPrecise = false
                     return true;
 
                 case GlobalAction.NextVolumeMeter:
-                    if (State.Value == Visibility.Visible)
-                        volumeMeters.SelectNext();
+                    if (State.Value != Visibility.Visible)
+                        return false;
+
+                    volumeMeters.SelectNext();
                     Show();
                     return true;
 
                 case GlobalAction.PreviousVolumeMeter:
-                    if (State.Value == Visibility.Visible)
-                        volumeMeters.SelectPrevious();
+                    if (State.Value != Visibility.Visible)
+                        return false;
+
+                    volumeMeters.SelectPrevious();
                     Show();
                     return true;
 

```

I think this is probably fine though.

OliBomby committed:
prevent volume control from eating inputs

OliBomby committed:
allow alt scroll for volume in editor

OliBomby commented:
> I'm not 100% sure about the auto scroll on single click.

I'll ask some mappers for feedback about this. I'm also not 100% sure.

Next to your patch I also disabled Alt+Scroll to seek, which I had previously enabled. We probably don't want to disable the ability to change volume in the editor with Alt+Scroll.

OliBomby committed:
Make hit object and sample point seek keybinds configurable

peppy commented:
I don't know what's going on here but it looks pretty broken:

https://github.com/user-attachments/assets/1c7a9a74-2e29-4f9f-a1c3-c27c9c4d3b93

Probably needs investigation before considering adding this UX behaviour.

OliBomby commented:
> I don't know what's going on here but it looks pretty broken:
> 
>  osu.2024-08-19.at.04.18.29.mp4 
> Probably needs investigation before considering adding this UX behaviour.

I fail to see what is broken in the video. It's supposed to seek to neighbouring hitobjects with Alt+left/right and seek to nearby sample control points with Alt+Shift+left/right. That seems to be happening in the video.

Joehuu commented:
Adding an arrow to the popover makes the bug more apparent. With this diff:
```diff
diff --git a/osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs b/osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs
index 731fe8ae6a..81d5b2e0cb 100644
--- a/osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs
+++ b/osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs
@@ -12,6 +12,7 @@
 using osu.Framework.Graphics;
 using osu.Framework.Graphics.Containers;
 using osu.Framework.Graphics.Cursor;
+using osu.Framework.Graphics.Shapes;
 using osu.Framework.Graphics.UserInterface;
 using osu.Framework.Input.Events;
 using osu.Game.Audio;
@@ -145,6 +146,13 @@ public partial class SampleEditPopover : OsuPopover
             [Resolved(canBeNull: true)]
             private EditorBeatmap beatmap { get; set; } = null!;
 
+            protected override Drawable CreateArrow() => new Triangle
+            {
+                Size = new Vector2(20),
+                Anchor = Anchor.Centre,
+                Origin = Anchor.TopCentre,
+            };
+
             public SampleEditPopover(HitObject hitObject)
             {
                 this.hitObject = hitObject;
```



https://github.com/user-attachments/assets/a7f0114a-b678-4c3b-a2ad-fd0c1178d13e

Sometimes the popover doesn't go to the seeked object. Also I feel like adding the arrow is a good UX improvement because before I wouldn't know what I'm editing.

OliBomby committed:
Merge remote-tracking branch 'upstream/master' into doubleclick

OliBomby commented:
Ah I see the problem now. It seems I already fixed this in https://github.com/ppy/osu/pull/28772 but this PR wasn't updated since.

OliBomby committed:
add arrow indicator

bdach commented:
The arrow key to move around thing seems pretty cool for whatever it's worth.

Adding the arrow to the popover though, not sure. Weird to be adding it only in one place. The reason why I didn't originally do this even though `Popover` supports it is that it doesn't work with shadows:

![image](https://github.com/user-attachments/assets/8420e8f3-d4cf-47db-aab6-a44d455c7800)

Note that the shadow draws _over_ the arrow.

Maybe we need selection state to show on the sample/velocity pieces themselves. Probably with something like a yellow border or whatever, to match blueprints. It's something that I had in mind before, in that it's kinda difficult to tell right now which pieces you're editing specifically at any given time, especially when multiple objects are selected at a time.

OliBomby committed:
Revert "add arrow indicator"

This reverts commit 1ce9e97fd45bb81f13a8e6a799af43d6342922af.

OliBomby commented:
> Adding the arrow to the popover though, not sure. Weird to be adding it only in one place. The reason why I didn't originally do this even though `Popover` supports it is that it doesn't work with shadows:
> 
> Note that the shadow draws _over_ the arrow. It's not super trivial to fix.
> 
> Maybe we need selection state to show on the sample/velocity pieces themselves. Probably with something like a yellow border or whatever, to match blueprints. It's something that I had in mind before, in that it's kinda difficult to tell right now which pieces you're editing specifically at any given time, especially when multiple objects are selected at a time.

Seems like those solutions are all not trivial, so I'll keep it out of this PR.

I think the yellow border on the sample pieces makes sense if they are also individually selectable like hitobjects in the timeline.

peppy commented:
I was also very tempted to add the speech bubble marker. I wouldn't mind if it was just applied here for now, but either or I guess.

peppy commented on osu.Game/Screens/Edit/Editor.cs:
```diff
@@ -751,6 +752,22 @@ public bool OnPressed(KeyBindingPressEvent<GlobalAction> e)
                     bottomBar.TestGameplayButton.TriggerClick();
                     return true;
 
+                case GlobalAction.EditorSeekToPreviousHitObject:
```
I feel like these should allow key repeat, feels better?

```diff
diff --git a/osu.Game/Screens/Edit/Editor.cs b/osu.Game/Screens/Edit/Editor.cs
index 2933c89cd8..355d724434 100644
--- a/osu.Game/Screens/Edit/Editor.cs
+++ b/osu.Game/Screens/Edit/Editor.cs
@@ -714,6 +714,26 @@ protected override bool OnScroll(ScrollEvent e)
 
         public bool OnPressed(KeyBindingPressEvent<GlobalAction> e)
         {
+            // Repeatable actions
+            switch (e.Action)
+            {
+                case GlobalAction.EditorSeekToPreviousHitObject:
+                    seekHitObject(-1);
+                    return true;
+
+                case GlobalAction.EditorSeekToNextHitObject:
+                    seekHitObject(1);
+                    return true;
+
+                case GlobalAction.EditorSeekToPreviousSamplePoint:
+                    seekSamplePoint(-1);
+                    return true;
+
+                case GlobalAction.EditorSeekToNextSamplePoint:
+                    seekSamplePoint(1);
+                    return true;
+            }
+
             if (e.Repeat)
                 return false;
 
@@ -751,26 +771,9 @@ public bool OnPressed(KeyBindingPressEvent<GlobalAction> e)
                 case GlobalAction.EditorTestGameplay:
                     bottomBar.TestGameplayButton.TriggerClick();
                     return true;
-
-                case GlobalAction.EditorSeekToPreviousHitObject:
-                    seekHitObject(-1);
-                    return true;
-
-                case GlobalAction.EditorSeekToNextHitObject:
-                    seekHitObject(1);
-                    return true;
-
-                case GlobalAction.EditorSeekToPreviousSamplePoint:
-                    seekSamplePoint(-1);
-                    return true;
-
-                case GlobalAction.EditorSeekToNextSamplePoint:
-                    seekSamplePoint(1);
-                    return true;
-
-                default:
-                    return false;
             }
+
+            return false;
         }
 
         public void OnReleased(KeyBindingReleaseEvent<GlobalAction> e)

```

peppy commented on osu.Game/Screens/Edit/Editor.cs:
```diff
@@ -1077,6 +1094,69 @@ private void seekControlPoint(int direction)
+        }
+
+        [CanBeNull]
+        public event Action<double> ShowSampleEditPopoverRequested;
```
This flow is a bit ugly. Probably fine maybe but still ugly.

Move `public` things to top of file and document though.

OliBomby committed:
allow repeating on seek actions

OliBomby committed:
move public member to top of file

peppy commented:
@OliBomby I'm thinking we should change this to be <kbd>Ctrl</kbd> instead of <kbd>Alt</kbd> to match user expectations (ie. matches traversing text by words in a text box). What say you?

peppy commented on osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs:
```diff
@@ -43,15 +49,35 @@ public SamplePointPiece(HitObject hitObject)
 
         protected override Color4 GetRepresentingColour(OsuColour colours) => AlternativeColor ? colours.Pink2 : colours.Pink1;
 
+        protected virtual double GetTime() => HitObject is IHasRepeats r ? HitObject.StartTime + r.Duration / r.SpanCount() / 2 : HitObject.StartTime;
```
I don't have much faith in this not running into some weird floating point issue when requesting display, but let's see how it goes.

OliBomby replied:
It seems that it can run into issues if the sample point time is not calculated exactly the same way. I've added some leniency just in case.

peppy commented on osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs:
```diff
@@ -43,15 +49,35 @@ public SamplePointPiece(HitObject hitObject)
 
         protected override bool OnClick(ClickEvent e)
         {
+            editorClock?.SeekSmoothlyTo(GetTime());
```
I'm still not sure about this UX. Maybe best to only trigger this seek when using the seek keys?

Results in big reach:

https://github.com/user-attachments/assets/9277306a-ae8e-44e9-96d0-1d50b94aff76

OliBomby replied:
I see your point. I'm not sure what you mean by the seek keys. Should it only seek when you click it while Control is held? 

I think it might be best to trigger the seek on a right click maybe, or some other mouse button, because I expect the mapper to be already using the hotkeys for assigning hitsounds when using this feature, so adding modifiers on click makes the flow more complicated.

OliBomby replied:
I would've liked to keep the right-click navigation but I assume that's a peppy-no-go

peppy replied:
I dunno. It's too many random UX behaviours. I want to avoid this unless it's well defined (ie. something that matches other apps, or something implemented everywhere in lazer).

It's a no-go for this PR.

OliBomby committed:
fix difference in sample time calculation

OliBomby committed:
Add float comparison leniency just in case

OliBomby committed:
change default seek hotkeys

OliBomby committed:
Automatic seek to sample point on right-click

peppy commented:
Can we not keep the ctrl-shift for sample points? I feel like we may want to keep alt for other usages in the future. Or leave it unused.

OliBomby committed:
change sample seek keybind to ctrl shift

OliBomby commented:
> Can we not keep the ctrl-shift for sample points? I feel like we may want to keep alt for other usages in the future. Or leave it unused.

That also works I guess.

peppy reviewed (changes requested):
As proposed.

OliBomby committed:
dont seek on right-click, only on keyboard request

OliBomby committed:
Merge remote-tracking branch 'upstream/master' into doubleclick

OliBomby committed:
fix test
